<?xml version="1.0"?>
<launch>
    <arg name="showrviz" default="true" />
    <!-- -->
    <arg name="usebag" default="false" />
    <arg name="sequence_py" default="00" />
    <arg name="weighttype" default="2" />
    <arg name="curvedfilter" default="1" />
    <arg name="groundfilter" default="1" />
    <arg name="featurePreExtract" default="1" />


    <param name="velodyne_points" type="string" value="/velodyne_points" />
    <param name="pfilter_input_cloud" type="string" value="/pfilter_input_cloud" />

    <group if="$(arg usebag)">
        <arg name="bag_filename" default="/home/r/dataset/kitti_velod_bags/07.bag" />
        <node pkg="rosbag" type="play" name="rosbag_play" args="--clock $(arg bag_filename)  " />
    </group>
    <group unless="$(arg usebag)">
        <include file="$(find kitti_tool)/launch/kitti.launch">
            <arg name="sequence_number" default="$(arg sequence_py)" />
        </include>
    </group>


    <node pkg="pfilter" type="curvedVoxel_node" name="curvedVoxel_node" output="screen" />


    <group unless="$(arg featurePreExtract)">
        <node pkg="pfilter" type="PFilter_laser_processing_node"
            name="PFilter_laser_processing_node"
            output="screen" />
    </group>


    <node pkg="pfilter" type="PFilter_odom_estimation_node" name="PFilter_odom_estimation_node"
        output="screen" />

    <!-- For Velodyne VLP-16 
    <param name="scan_line" value="16" />
    -->

    <!-- For Velodyne HDL-32  -->
    <!-- <param name="scan_line" value="32" /> -->


    <!-- For Velodyne HDL-64 -->
    <param name="scan_line" value="64" />
    <arg name="sensorFrameId" default="base_link" />

    <!---
    Sim Time -->
    <param name="/use_sim_time" value="false" />
    <param name="scan_period" value="0.1" />
    <param name="vertical_angle" type="double" value="2.0" />
    <param name="max_dis" type="double" value="90.0" />
    <param name="min_dis" type="double" value="3.0" />


    <param name="k_new" type="string" value="0" />
    <param name="theta_p" type="string" value="0.4" />
    <param name="theta_max" type="string" value="75" />
    <param name="map_resolution" type="double" value="0.4" />
    <param name="yamlConfigFile" type="string"
        value="/home/r/catkin_wss/pfilter_ws/src/PFilter-noetic/config/config.yaml" />
    <param name="sensorFrameId" type="string" value="$(arg sensorFrameId)" />

    <param name="weightType" type="int" value="$(arg weighttype)" />
    <param name="curvedfilter" type="int" value="$(arg curvedfilter)" />
    <param name="groundfilter" type="int" value="$(arg groundfilter)" />
    <param name="featurePreExtract" type="int" value="$(arg featurePreExtract)" />
    <!---       
-->
    <node pkg="tf" type="static_transform_publisher" name="word2map_tf"
        args="0 0 0 0 0 0 /world /map 10" />

    <group if="$(arg showrviz)">
        <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz"
            args="-d $(find pfilter)/rviz/pfilter.rviz" />
    </group>


    <node pkg="hector_trajectory_server" type="hector_trajectory_server"
        name="trajectory_server_loam" ns="gt">
        <param name="/target_frame_name" value="world" />
        <param name="/source_frame_name" value="$(arg sensorFrameId)" />
        <param name="/trajectory_update_rate" value="10.0" />
        <param name="/trajectory_publish_rate" value="10.0" />
    </node>
    <node pkg="hector_trajectory_server" type="hector_trajectory_server"
        name="trajectory_server_loam" ns="$(arg sensorFrameId)">
        <param name="/target_frame_name" value="world" />
        <param name="/source_frame_name" value="$(arg sensorFrameId)" />
        <param name="/trajectory_update_rate" value="10.0" />
        <param name="/trajectory_publish_rate" value="10.0" />
    </node>

</launch>